### PR TITLE
Progress report

### LINKED ISSUE #321: Progress report
On scripts that take a longer time to execute it would be nice to have some kind of progress report (e.g. script that needs to process 100K files or something similar).

Suggestion is to implement interface like this:

```csharp
ProgressReport.Report(double percentageDone); //  where percentageDone is between 0.0 and 1.0
```

Progress can be displayed next to run button.

@aelij I could work on this and submit a PR if you agreed with the suggestions.

### PR DESCRIPTION
**Pull request by stanac**
@aelij thank you for such a detailed explanation of the inner workings of the app.

Please let me know if you want me to change something, or do something differently. Below is a screen recording of the progress bar.

progress bar

Fixes #321

### CODE REVIEW

File: src/RoslynPad.Common.UI/ViewModels/OpenDocumentViewModel.cs [hunk 1]
```diff
@@ -872,6 +877,21 @@ public bool IsDirty
...
+            {
+                if (_reportedProgress != value)
+                {
+                    SetProperty(ref _reportedProgress, value);
+                    OnPropertyChanged(nameof(ReportedProgressVisible));
+                }
+            }
+        }
+
+        public bool ReportedProgressVisible => ReportedProgress.HasValue;
```
Comment (aelij) [inline src/RoslynPad.Common.UI/ViewModels/OpenDocumentViewModel.cs]: Rename to `HasReportedProgress`

File: src/RoslynPad.Runtime/Runtime/ProgressReport.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿namespace RoslynPad.Runtime
+{
+    /// <summary>
+    /// Reports progress to RoslynPad GUI
+    /// </summary>
+    public static class ProgressReport
```
Comment (aelij) [inline src/RoslynPad.Runtime/Runtime/ProgressReport.cs]: Remove this class. I'd like a single entry point for RP APIs, and that's `Helpers`.

File: src/RoslynPad.Common.UI/ViewModels/OpenDocumentViewModel.cs (hunk 1 above)
Comment (stanac) [reply src/RoslynPad.Common.UI/ViewModels/OpenDocumentViewModel.cs]: renamed

File: src/RoslynPad.Runtime/Runtime/ProgressReport.cs (hunk 2 above)
Comment (stanac) [reply src/RoslynPad.Runtime/Runtime/ProgressReport.cs]: removed

**Comment by stanac** [comment]
Changes are added to PR, also previously I have forgot about Avalonia UI, which now also supports progress report with a minor issue. ProgressBar on Avalonia UI does not respect width (or it's possibly a bug on their side, so it's somewhat wider than on WPF UI).

**Comment by Kirides** [comment]
Does this support Taskbar progress in Windows? / Do you plan to integrate Windows' progress Taskbar integration?

i would love such a feature. I mainly use RoslynPad as a script runner and have a whole bunch of scripts that would benefit from progress reporting.

**Comment by aelij** [comment]
@Kirides this could be a nice addition, but in a different PR. Btw you don't need `Microsoft-WindowsAPICodePack`, it's built into WPF (see https://docs.microsoft.com/en-us/dotnet/api/system.windows.shell.taskbariteminfo)

**Comment by aelij** [comment]
Thanks for the contribution, @stanac 🎉

**Pull request merged**